src/core/public/http/fetch.ts

```diff
@@ -95,38 +102,53 @@ export class Fetch {
     });
   };
 
-  private createRequest(path: string, options?: HttpFetchOptions): Request {
+  private createRequest(options: HttpFetchOptionsWithPath): Request {
     // Merge and destructure options out that are not applicable to the Fetch API.
-    const { query, prependBasePath: shouldPrependBasePath, asResponse, ...fetchOptions } = merge(
+    const {
+      query,
+      prependBasePath: shouldPrependBasePath,
+      asResponse,
+      asSystemRequest,
+      ...fetchOptions
+    } = merge(
       {
         method: 'GET',
         credentials: 'same-origin',
         prependBasePath: true,
+      },
+      options,
+      {
         headers: {
-          'kbn-version': this.params.kibanaVersion,
           'Content-Type': 'application/json',
+          ...options.headers,
+          'kbn-version': this.params.kibanaVersion,
         },
-      },
-      options || {}
+      }
     );
     const url = format({
-      pathname: shouldPrependBasePath ? this.params.basePath.prepend(path) : path,
+      pathname: shouldPrependBasePath ? this.params.basePath.prepend(options.path) : options.path,
       query,
     });
 
-    if (
-      options &&
-      options.headers &&
-      'Content-Type' in options.headers &&
-      options.headers['Content-Type'] === undefined
-    ) {
-      delete fetchOptions.headers['Content-Type'];
+    // Make sure the system request header is only present if `asSystemRequest` is true.
+    if (asSystemRequest) {
+      fetchOptions.headers['kbn-system-request'] = 'true';
+    } else {
+      if ('kbn-system-request' in fetchOptions.headers) {
+        Reflect.deleteProperty(fetchOptions.headers, 'kbn-system-request');
```

Comment:
That would break the plugin code. Shouldn't we warn a plugin about the preferable solution with `{asSystemRequest: true}` instead?

Reply:
What do you think about making this throw if any `kbn-*` headers are added? I think it would be safe to consider those "internal" headers that are not allowed to be used by plugins.

Reply:
My goal here is to prevent the behavior details of these leaking through to plugins.

Reply:
> What do you think about making this throw if any kbn-* headers are added?

works for me